### ISSUE TITLE
setup troubles [support]

### ISSUE THREAD

**Issue by eMPee584** [lucaslorentz/caddy-docker-proxy#72]
after failing to set up nginx-proxy with VIRTUAL_PATH, I tried deploying your docker image, but can't make it work.
There's a mattermost install on the server instantiated from https://github.com/mattermost/mattermost-docker.git with a `docker-compose.override.yml`:

```
version: "2"

services:
  web:
    build: web
    ports:
      - "8580:80"
      - "8543:443"
    labels:
      - caddy.address=tii.tu-dresden.de:80/mm
      - caddy.targetport=8580
      - caddy.targetpath=/
      - caddy.caddy.targetprotocol=http

```
but if I go to that address — nothing. How can I even debug the problem? `docker exec -it {/caddy} bash` doesn't work, the container log just tells me
```
Activating privacy features... done.
http://:2015
```
strange thing is, `lsof -i` on the docker host doesn't even show that caddy has any port opened? .. :thinking:

**Comment by lucaslorentz** [comment]
**Debugging problems**
Use `-log stdout` parameter, like: https://github.com/lucaslorentz/caddy-docker-proxy/blob/master/examples/service-proxy.yaml#L16.
Then you can do `docker container logs CADDYCONTAINERID` or `docker service logs CADDYSERVICEID` and view the logs.

**Command line args**
It is recommended to provide at least those parameters to caddy-docker proxy:
```
command: -email email@example.com -agree=true -log stdout
```

**Ports**
Docker ports binds host ports to container ports. Because caddy-docker-proxy is also inside docker, it does container to container communication. So you need to use `caddy.targetport=80` instead of `caddy.targetport=8580`.

**Bash inside caddy-docker-proxy**
We have multiple kind of images, the default image `lucaslorentz/caddy-docker-proxy` is a very small image, and doesn't have sh or bash inside it. If you want to run bash inside the container to diagnose something like pinging or curl another container, use image `lucaslorentz/caddy-docker-proxy:alpine`.

**Fixes to labels**
Please, remove label `caddy.caddy.targetprotocol` key is wrong and by default it already uses http.
Remove label `caddy.targetpath` as well, by default it uses /.

**Test**
```
docker run -it --rm -p 80:80 -p 443:443 -v //var/run/docker.sock://var/run/docker.sock lucaslorentz/caddy-docker-proxy -email email@example.com -agree=true -log stdout
```
```
docker run -it --rm -l caddy.address=tii.tu-dresden.de:80/mm -l caddy.targetport=8000 jwilder/whoami
```
```
curl -H Host:tii.tu-dresden.de http://localhost/mm
```
Output:
I'm 4ea7573e6862

**Getting HTTPS**
After you get things working, you can try to enable https by changing your address label to:
```
- caddy.address=tii.tu-dresden.de/mm
```

**Comment by eMPee584** [comment]
Ok I have a hard time following your explanations because I am new to the whole container business, but I am trying.
So the yaml files are supposed to be used with `docker-compose up -f`.
I tried before without the targetpath and protocol lines..
The thing is, I do not want to deploy all the containers from one file; I want the caddy proxy pick them up one by one. The whoami entries in the yaml suggest all applications should be deployed together with this.. Do I have to use such a yaml at all? I thought it is enough to pull the docker image from the repo and add the labels to my application containers... :confused: 

Basically, I want webapp containers listening on port xy80/xy443 to appear at /foo.

**Comment by eMPee584** [comment]
ok your example commands are helpful.. i might even figure this out :rofl:

**Comment by eMPee584** [comment]
hmm lots of `# Container e96bddf15071057a9f18e5cccc8b02e5230e83c8d5d12afeb65c4450b6a268cd and caddy are not in same network
`

**Comment by eMPee584** [comment]
ok so I tried to put all app containers into the bridge network with by adding `    network_mode: "bridge"
` to each service in my mattermost/docker-compose.override.yml but that broke the apps discovery.
How can I make caddy-docker-proxy listen in on other networks?

**Comment by lucaslorentz** [comment]
> So the yaml files are supposed to be used with docker-compose up -f.

To be honest, I don't use docker-compose, I use docker stacks to deploy yaml files.
I would expect docker-compose to become deprecated.
First make sure you have a swarm cluster, even if it is a cluster of 1 node :-)
```
docker swarm init
```

Then deploy your yaml file to a stack with:
```
docker stack deploy -c your-file.yaml stackname
```

> The thing is, I do not want to deploy all the containers from one file; I want the caddy proxy pick them up one by one. The whoami entries in the yaml suggest all applications should be deployed together with this.. Do I have to use such a yaml at all? I thought it is enough to pull the docker image from the repo and add the labels to my application containers... 😕

It's not required to use yaml. The 3 commands I provided above just work.
But I still recommend using swarm orchestrator to ensure your containers are always running.
You could use just pure command line to create swarm services, but I think it's better to prepare your deployments as yaml, so you can update it and reapply the yaml as many times as you want.

I have production environments with several yaml files, each one with some services.
Sorry for the misleading example file, I put everything together just for simplicity.

One important thing is that once you break your yaml file into multiples and deploy each one to a stack, you need to create the network separately, because the network needs to be shared between all stacks.

Create a network with:
```
docker network create caddy -d overlay
```

Then on each yaml file you have, add the network as external network and link it to your services (both apps services and caddy service):
```
services:
  caddy:
    image: lucaslorentz/caddy-docker-proxy
    networks:
      - caddy
    ...

networks:
  caddy:
    external: true
```

> hmm lots of # Container e96bddf15071057a9f18e5cccc8b02e5230e83c8d5d12afeb65c4450b6a268cd and caddy are not in same network
> ok so I tried to put all app containers into the bridge network with by adding network_mode: "bridge" to each service in my mattermost/docker-compose.override.yml but that broke the apps discovery.
How can I make caddy-docker-proxy listen in on other networks?

About the network, it doesn't really matter if it is bridge or overlay or other drivers. But caddy must be able to reach the app containers, so they need to be on same network.

Caddy sees all containers, but it ignores the ones it cannot reach due to not being on the same network.

**Comment by eMPee584** [comment]
mhh thanks..
in the meantime, I am having a looking at traefik .. which supposedly is built for this use case from the ground up.

**Comment by eMPee584** [comment]
doing containers right is not as easy as I hoped :grin:

**Comment by lucaslorentz** [comment]
Glad you found your way.
Traefik is great. And it's very well documented.

**Comment by eMPee584** [comment]
not yet :rofl: 
docker ecosystem is buzzing with great projects - just needs some time to find a way through the jungle :sunglasses:
